content/courses/advanced-ada/chapters/subprograms.rst

```diff
@@ -6,21 +6,456 @@ Subprograms
+:ada:`Is_Zero` function because :ada:`(I = 0)` is the actual implementation of
+the function. Note that this is only possible for expression functions; you
+cannot have a function implemented with a block in a package specification. For
+example, the following code is wrong and won't compile:
+
+.. code:: ada manual_chop compile_button project=Courses.Advanced_Ada.Numerics.Simple_Expression_Function_3
+    :class: ada-nocheck
+
+    !expr_func.ads
+    package Expr_Func is
+
+       function Is_Zero (I : Integer) return Boolean is
+       begin
+          return I = 0;
+       end Is_Zero;
+
+    end Expr_Func;
+
+We can, of course, separate the function declaration from its implementation as
+an expression function. For example, we can rewrite the first version of the
+:ada:`Expr_Func` package and move the expression function to the body of the
+package:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Numerics.Simple_Expression_Function_4
+
+    package Expr_Func is
+
+       function Is_Zero (I : Integer) return Boolean;
+
+    end Expr_Func;
+
+    package body Expr_Func is
+
+       function Is_Zero (I : Integer) return Boolean is
+         (I = 0);
+
+    end Expr_Func;
+
+In addition, we can use expression functions in the private part of a
+package specification. For example, the following code declares the
+:ada:`Is_Valid` function in the specification of the :ada:`My_Data` package,
+while its implementation is an expression function in the private part of the
+package specification:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Numerics.Private_Expression_Function_1
+
+    package My_Data is
+
+       type Data is private;
+
+       function Is_Valid (D : Data) return Boolean;
+
+    private
+
+       type Data is record
+          Valid : Boolean;
+       end record;
+
+       function Is_Valid (D : Data) return Boolean is
+          (D.Valid);
+
+    end My_Data;
+
+Naturally, we could write the function implementation in the package body
+instead:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Numerics.Private_Expression_Function_2
+
+    package My_Data is
+
+       type Data is private;
+
+       function Is_Valid (D : Data) return Boolean;
+
+    private
+
+       type Data is record
+          Valid : Boolean;
+       end record;
+
+    end My_Data;
+
+    package body My_Data is
+
+       function Is_Valid (D : Data) return Boolean is
+          (D.Valid);
+
+    end My_Data;
+
+
+Conditional Expressions
+-----------------------
+
+In the previous section, we've seen simple expressions such as :ada:`I = 0` or
+:ada:`D.Valid`. A conditional expression, as the name implies, is an
+expression that contains a condition. This might be an "if-expression" (in the
+:ada:`if ... then ... else` form) or a "case-expression" (in the
+:ada:`case ... is when =>` form). The result of a conditional expression is
+always a Boolean value.
```

Comment:
"The result of a conditional expression is always a Boolean value."  That's not right, if I understand the intent. The result is a value of a type determined by the context. For example, the result of the conditional expression implementing function Max is Integer, because that's the type that Max is declared to return. The values of A and B are checked to ensure they are Integer values.

Reply:
You're absolutely right, of course! Even my example contradicts this statement, so I'll just remove it.

(I'm not sure how this ended up here, but I guess this is just a copy-and-paste error, as this phrase is part of the next subsection on quantified expressions.)